presto-main/src/main/java/com/facebook/presto/operator/LookupJoinPageBuilder.java

```diff
@@ -0,0 +1,141 @@
+    private final Set<Integer> probeIndices = new HashSet<>();
+    private final PageBuilder buildPageBuilder;
+    private final int buildOutputChannelCount;
+    private int probeBlockBytes;
+
+    public LookupJoinPageBuilder(List<Type> buildTypes)
+    {
+        this.buildPageBuilder = new PageBuilder(buildTypes);
+        this.buildOutputChannelCount = buildTypes.size();
+    }
+
+    public boolean isFull()
+    {
+        return probeBlockBytes + buildPageBuilder.getSizeInBytes() >= DEFAULT_MAX_PAGE_SIZE_IN_BYTES || buildPageBuilder.isFull();
+    }
+
+    public boolean isEmpty()
+    {
+        return probeIndexBuilder.isEmpty() && buildPageBuilder.isEmpty();
+    }
+
+    public void reset()
+    {
+        probeIndexBuilder.clear();
+        probeIndices.clear();
+        buildPageBuilder.reset();
+        probeBlockBytes = 0;
+    }
+
+    /**
+     * append the index for the probe and copy the row for the build
+     */
+    public void appendRow(JoinProbe probe, LookupSource lookupSource, long joinPosition)
+    {
+        buildPageBuilder.declarePosition();
+        // write probe columns
+        appendProbeIndex(probe);
+        // write build columns
+        lookupSource.appendTo(joinPosition, buildPageBuilder, 0);
+    }
+
+    /**
+     * append the index for the probe and append nulls for the build
+     */
+    public void appendNull(JoinProbe probe, LookupSource lookupSource)
+    {
+        buildPageBuilder.declarePosition();
+        // write probe columns
+        appendProbeIndex(probe);
+        // write nulls into build columns
+        for (int i = 0; i < lookupSource.getChannelCount(); i++) {
+            buildPageBuilder.getBlockBuilder(i).appendNull();
+        }
+    }
+
+    public Page build(JoinProbe probe)
+    {
+        Block[] blocks = new Block[probe.getOutputChannelCount() + buildOutputChannelCount];
+        int[] probeIndices = probeIndexBuilder.toIntArray();
+        for (int i = 0; i < probe.getOutputChannelCount(); i++) {
+            Block block = probe.getPage().getBlock(probe.getOutputChannels()[i]);
+            if (block instanceof LazyBlock) {
+                // TODO: on one hand, we shouldn't load the block at this point; on the other, we cannot wrap a lazy block inside a dictionary
+                block = ((LazyBlock) block).getBlock();
+            }
+            blocks[i] = block instanceof DictionaryBlock ? ((DictionaryBlock) block).mask(probeIndices) : new DictionaryBlock(block, probeIndices);
+        }
+
+        int[] dictionaryIds = new int[buildPageBuilder.getPositionCount()];
+        for (int i = 0; i < buildPageBuilder.getPositionCount(); i++) {
+            dictionaryIds[i] = i;
+        }
+        for (int i = 0; i < buildOutputChannelCount; i++) {
+            // make sure all the blocks are dictionary blocks
+            blocks[probe.getOutputChannelCount() + i] = new DictionaryBlock(buildPageBuilder.getBlockBuilder(i).build(), dictionaryIds);
```

Comment:
* I know `probe.getOutputChannelCount()` will surely be inlined, but assigning it locally will improve readability.

> `buildPageBuilder.getBlockBuilder(i).build()`

I think it would be safer to `buildPageBuilder.build()` and operate on returned page's blocks. `PageBuilder#build` has some validation. Second, it could change internal state.